### PR TITLE
Fixes [SR-78] swift compiler seg fault

### PR DESCRIPTION
**Pull request by manavgabhawala**
Fixes [SR-78](https://bugs.swift.org/browse/SR-78) swift compiler segmentation fault. The IRGen module would crash when a generic (empty) struct was part of the source code because it tried to call std::next on an end iterator, adds in a guard to prevent against that and also adds a test case to add in this check.

Added test case that caused the old compiler to crash (I'm not sure if this is the right way to integrate it in the testing suite).

### CODE REVIEW

**Comment by nadavrot** [comment]
@manavgabhawala I prefer not to merge this change because it has a bunch of changes to different files. Some of the commit messages are not informative, like "Update Heap.cpp", twice.

Please submit a new pull request with a single commit that includes a fix for a single issue with a testcase.

File: lib/IRGen/GenType.cpp [hunk 1]
```diff
@@ -2032,6 +2032,10 @@ SILType irgen::getSingletonAggregateFieldType(IRGenModule &IGM,
     // If there's only one stored property, we have the layout of its field.
     auto allFields = structDecl->getStoredProperties();
     auto field = allFields.begin();
```
Comment (AlexDenisov) [inline lib/IRGen/GenType.cpp]: Wouldn't it be better to check for empty here?
```
if (allFields.empty()) {
  return SILType();
}
```

**Comment by manavgabhawala** [comment]
Sorry about the multiple commits, I was trying to get my feet wet. I created a new clean pull request and included the suggestion by @AlexDenisov. Moved to #293

File: lib/IRGen/GenType.cpp (hunk 1 above)
Comment (manputra) [inline lib/IRGen/GenType.cpp]: Yes

**Pull request closed without merging**